test/datatype.test.js

```diff
@@ -52,6 +52,36 @@ describe('datatypes', function() {
     });
     Account.definition.properties.item.type.should.not.equal(String);
   });
+  it('should resolve array prop with connector specific metadata', function() {
+    const model = db.define('test', {
+      randomReview: {
+        type: [String],
+        mongodb: {
+          dataType: 'Decimal128',
+        },
+      },
+    });
+    model.definition.properties.randomReview.type.should.deepEqual(Array(String));
+    model.definition.properties.randomReview.mongodb.should.deepEqual({dataType: 'Decimal128'});
+  });
+
+  it('should coerce array of primitive types from string', async () => {
+    const dateArrayModel = db.define('dateArrayModel', {
+      bunchOfDates: [
+        {
+          type: Date,
+        },
+      ],
+    });
+    const dateVal = new Date('2019-02-21T12:00:00').toISOString();
+    const created = await dateArrayModel.create({
+      bunchOfDates: [dateVal,
+        dateVal,
+        dateVal],
+    });
+    created.bunchOfDates[0].should.be.an.instanceOf(Date);
+    created.bunchOfDates[0].should.equal(Date(dateVal));
+  });
 
```

Comment:
I think it would be good to test coercion for array types defined using the approach shown in the first test too.

```ts
const dateArrayModel = db.define('dateArrayModel', {
  bunchOfDates: {
    type: [Date],
  },
});

// ...
created.bunchOfDates[0].should.be.an.instanceOf(Date);
created.bunchOfDates[0].should.equal(Date(dateVal));
```     

Unless that's something we have already covered by other tests?

Similarly, test how connector-specific properties are preserved for the second flavor.

```ts
const model = db.define('model', {
   bunchOfDates: [
     {
       type: Date,
        // is this the right syntax?
        mongodb: {
           dataType: 'Decimal128',
        },  
     },
   ],
 });
```

Reply:
Okay I can try this out as well, I'll check if there is existing coverage.

Reply:
>I think it would be good to test coercion for array types defined using the approach shown in the first test too.
Agreed, will add the test case in my next commit.
> Similarly, test how connector-specific properties are preserved for the second flavor.
Should we allow this? I don't think we handle this case and this was what I tried to fix with my first commit, but I feel like it can be addressed properly in a wider context and we should advocate our users use first flavour (`prop: {type:[String], mongodb: ...}`. FYI this is how we interpret that definition:
def:
```
model.definition.properties.customMongoProp2
Object {0: Object, type: Array(1)}
should:Assertion
type:Array(1) []
0:Object {type: , mongodb: Object}
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
```
result:
```
model.definition.properties.customMongoProp2
Object {0: Object, type: Array(1)}
should:Assertion
type:Array(1) []
0:Object {type: , mongodb: Object}
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
```

Reply:
Just followed the discussion above, I think the definition is missing an object property name:

```
const model = db.define('model', {
   bunchOfDates: [
     {
       // Please note the object should contain each property name and its definition as a key-value pair
       date: {
         type: Date,
         // is this the right syntax?
         mongodb: {
           dataType: 'Decimal128',
         },  
       }
     },
   ],
 });
```

Reply:
FYI This is a list of nested property definitions I've tested with loopback-connector-mongodb@3.x:
https://github.com/strongloop/loopback-connector-mongodb/blob/3.x/docs/decimal128.md

Reply:
Thanks Janny. I think the use cases you pointed out are covered in the PR I created for mongodb (see https://github.com/strongloop/loopback-connector-mongodb/pull/501). We are looking at specifying an array type like `{"emails": [{"type": "string", "length": 64}]}` (see https://loopback.io/doc/en/lb3/LoopBack-types.html#array-types). I see the same results for even non-connector specific properties, so I think it's possibly a bug or we support it and maybe I need to change my assertions:

```
model.definition.properties.emails
Object {0: Object, type: Array(1)}
should:Assertion
type:Array(1) []
0:Object {type: "string", length: 64}
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
```
I'm going to leave it out of this PR since the fix in this PR is for a) making sure Date array properties are properly coerced, and b) making sure we coerce primitive types on update operation.